### PR TITLE
[Thumbnails] Add xxh32 hash of config to filename

### LINKED ISSUE #12424: [Improvement]: use a hash of the thumbnail's input in the target filename
## What

1. hash the original image (say, with MD5)
2. hash the thumbnail config (the relevant parts)
3. hash any other fact or input which could cause the final thumbnail bytes to change (crop guides)
4. use all these and create a thumbnail hash part of which (say, first 8 chars, like Git commits) are used in the final thumbnail filename

## Why?
- it makes the thumbnail name represent the sum of all its inputs, meaning that **if the thumbnail exists** at the expected filename, **it's up to date**, if it doesn't exist, it's not, we need to create it
- it automatically creates a cache buster directly in the filename
- it allows us to keep multiple copies of the same thumbnail and optionally delete the orphan ones at a later point. This also means that, if we change the input (original image or thumbnail config) and then revert it, we don't need to regenerate because the old thumbnail already matches the target (we're exchanging storage space for compute resources, storage is always cheaper)

See https://github.com/pimcore/pimcore/issues/12057#issuecomment-1154908126

### PR DESCRIPTION
**Pull request by brusch**
Fixes #12424

### CODE REVIEW

File: models/Asset/Image/Thumbnail/Processor.php [hunk 1]
```diff
@@ -164,7 +164,15 @@ public static function process(Asset $asset, Config $config, mixed $fileSystemPa
             $fileExtension = 'jpg';
         }
 
-        $filename .= '.' . $fileExtension;
+        $configHash = hash("crc32b", serialize([
```
Comment (jdreesen) [inline models/Asset/Image/Thumbnail/Processor.php]: Why not use something more modern like [`xxHash`](https://php.watch/versions/8.1/xxHash)?

```suggestion
        $configHash = hash('xxh3', serialize([
```

Comment (brusch) [reply models/Asset/Image/Thumbnail/Processor.php]: Yeah, why not, but probably `xxh32` to keep the hash shorter.

**Comment by brusch** [comment]
@dvesh3 ready 4 review now 😊 thx

File: models/Asset/Image/Thumbnail/Processor.php [hunk 2]
```diff
@@ -164,7 +164,7 @@ public static function process(Asset $asset, Config $config, mixed $fileSystemPa
             $fileExtension = 'jpg';
         }
 
-        $filename .= '.' . $fileExtension;
+        $filename .= '.' . $config->getHash() . '.'. $fileExtension;
```
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Processor.php:167]: Should we use a different separator for appending hashes? as dot `.` might cause issues with file extension detection.

Comment (jdreesen) [reply models/Asset/Image/Thumbnail/Processor.php:167]: Why? File extension is the part after the last `.`.

Comment (NiklasBr) [reply models/Asset/Image/Thumbnail/Processor.php:167]: Agree with @jdreesen, `$ext = pathinfo($filename, PATHINFO_EXTENSION)` works great and I have never seen it confused by things like excessive punctuation.

**Comment by brusch** [comment]
Resolved conflicts 😊

**Comment by dkarlovi** [comment]
I don't think this closes the original issue since it's (IIUC) hashing config only. Ideally, the hash would represent both the original file and the thumbnail config, meaning if either changes, the hash becomes stale.

I guess https://github.com/pimcore/pimcore/issues/13362 would be a prerequisite.

**Comment by brusch** [comment]
@dkarlovi you're right, this doesn't cover the original file, could you maybe provide a PR for https://github.com/pimcore/pimcore/issues/13362?

**Comment by dkarlovi** [comment]
Will try to do it later this week. :+1:

**Comment by brusch** [comment]
@dkarlovi created a follow-up to this PR, that allows to add custom criteria to the thumb config hash, so that we can do something like `$config->getHash([$asset->getHash()])` as soon as we have the hash stored on the asset ... see #14629

**Comment by brusch** [comment]
@dkarlovi just a quick reminder regarding https://github.com/pimcore/pimcore/issues/13362  😉  Would be already quite urgent as we're in the process of finalizing v11 😊

**Comment by dkarlovi** [comment]
@brusch sadly I'll not have time now to do it quickly, it'll have to wait for 11.1.

**Comment by brusch** [comment]
As this is kind of a breaking change, I'd definitely see this in 11.0.
We'll try to do handle that in our team then.

**Pull request merged**